server/src/main/java/io/crate/expression/reference/doc/lucene/StoredRowLookup.java

```diff
@@ -22,96 +22,295 @@
+            }
+        }
     }
 
-    private Map<String, Object> parseSource() {
-        return sourceParser.parse(fieldsVisitor.source());
+    private static class ColumnAndStoredRowLookup extends StoredRowLookup {
+
+        private record ColumnExpression(LuceneCollectorExpression<?> expression, ColumnIdent ident, String storageName) {}
+
+        private final List<ColumnExpression> expressions = new ArrayList<>();
+        private final ColumnFieldVisitor fieldsVisitor;
+
+        private ColumnAndStoredRowLookup(DocTableInfo table, String indexName, List<Symbol> columns) {
+            super(table, indexName);
+            this.fieldsVisitor = new ColumnFieldVisitor(table);
+            register(columns);
+        }
+
+        @Override
+        protected void registerRef(Reference ref) {
+            registerRef(ref, false);
+        }
+
+        private void registerRef(Reference ref, boolean fromParents) {
+            if (fromParents == false) {
+                // if we are inside an array, or an object type with ignored fields, then we need to load
+                // the relevant parent from stored fields
+                var storedParent = table.findParentReferenceMatching(ref, r ->
+                    r.valueType() instanceof ObjectType && r.columnPolicy() == ColumnPolicy.IGNORED ||
+                        r.valueType() instanceof ArrayType<?>);
+                if (storedParent != null) {
+                    this.fieldsVisitor.registerRef(storedParent);
+                }
+            }
+            if (ref.hasColumn(SysColumns.DOC) || ref.hasColumn(SysColumns.RAW)) {
+                // top-level _doc - we register all table columns
+                registerAll();
+            } else if (ref.valueType() instanceof GeoPointType) {
+                // geopoint always retrieved from stored fields
+                this.fieldsVisitor.registerRef(ref);
+            } else if (ref.valueType() instanceof ArrayType<?>) {
+                // arrays always retrieved from stored fields
+                this.fieldsVisitor.registerRef(ref);
+            } else if (ref.valueType() instanceof ObjectType) {
+                this.fieldsVisitor.registerRef(ref);
+                for (var leaf : table.getChildReferences(ref)) {
+                    registerRef(leaf, true);
+                }
+            } else {
+                LuceneCollectorExpression<?> expr = LuceneReferenceResolver.typeSpecializedExpression(ref);
+                if (expr instanceof DocCollectorExpression<?>) {
+                    this.fieldsVisitor.registerRef(ref);
+                } else {
+                    var column = ref.toColumn();
+                    if (column.isRoot() == false && column.name().equals(SysColumns.Names.DOC)) {
+                        column = column.shiftRight();
+                    }
+                    expressions.add(new ColumnExpression(expr, column, ref.storageIdent()));
+                }
+            }
```

Comment:
Similar here - I think the type's StorageSupport should have info on what kind of lookup is necessary instead of doing instanceof checks here